src/vanilla.ts

```diff
@@ -104,50 +80,34 @@ function createStore<
 
   const getState: GetState<TState> = () => state
 
-  const subscribeWithSelector = <StateSlice>(
-    listener: StateSliceListener<StateSlice>,
-    selector: StateSelector<TState, StateSlice> = getState as any,
-    equalityFn: EqualityChecker<StateSlice> = Object.is
-  ) => {
-    console.warn('[DEPRECATED] Please use `subscribeWithSelector` middleware')
-    let currentSlice: StateSlice = selector(state)
-    function listenerToAdd() {
-      const nextSlice = selector(state)
-      if (!equalityFn(currentSlice, nextSlice)) {
-        const previousSlice = currentSlice
-        listener((currentSlice = nextSlice), previousSlice)
-      }
-    }
-    listeners.add(listenerToAdd)
-    // Unsubscribe
-    return () => listeners.delete(listenerToAdd)
-  }
-
-  const subscribe: Subscribe<TState> = <StateSlice>(
-    listener: StateListener<TState> | StateSliceListener<StateSlice>,
-    selector?: StateSelector<TState, StateSlice>,
-    equalityFn?: EqualityChecker<StateSlice>
-  ) => {
-    if (selector || equalityFn) {
-      return subscribeWithSelector(
-        listener as StateSliceListener<StateSlice>,
-        selector,
-        equalityFn
-      )
-    }
-    listeners.add(listener as StateListener<TState>)
+  const subscribe: Subscribe<TState> = (listener: StateListener<TState>) => {
+    listeners.add(listener)
     // Unsubscribe
-    return () => listeners.delete(listener as StateListener<TState>)
+    return () => listeners.delete(listener)
   }
 
   const destroy: Destroy = () => listeners.clear()
   const api = { setState, getState, subscribe, destroy }
-  state = createState(
-    setState as unknown as CustomSetState,
-    getState as unknown as CustomGetState,
-    api as unknown as CustomStoreApi
+  state = (createState as PopArgument<typeof createState>)(
+    setState,
+    getState,
+    api
   )
-  return api as unknown as CustomStoreApi
+  return api as any
 }
 
-export default createStore
+type PopArgument<T extends (...a: never[]) => unknown> = T extends (
+  ...a: [...infer A, infer _]
+) => infer R
+  ? (...a: A) => R
+  : never
+
+export default create
+
+type CreateWithState = <T extends State>() => <
+  Mos extends [StoreMutatorIdentifier, unknown][] = []
+>(
+  initializer: StateCreator<T, [], Mos>
+) => Mutate<StoreApi<T>, Mos>
+
+export const createWithState: CreateWithState = () => create
```

Comment:
This looks like the trick of the whole thing. Is this internal use only? We don't want to expose this to library users.

Reply:
This is what would be the recommended way to create a store for typescript users. See [this comment](https://github.com/pmndrs/zustand/pull/662#discussion_r774789318).

Reply:
Why can't this be the default, then?
This is very important for zustand. Zustand really prefers to export one function.
It's unacceptable to export two functions from vanilla.ts.

Reply:
I think you have already considered, but can't we overload functions?

Reply:
I actually didn't think of overloads, that's a good idea, I'll update to use overloads.

Reply:
This is a major blocker. I hope it will work well!